torch/_functorch/aot_autograd.py

```diff
@@ -2819,5 +2819,170 @@ def forward(*runtime_args):
     return forward
 
 
+def aot_module_export(
+    mod: nn.Module,
+    args,
+    decompositions: Optional[Dict] = None,
+    keep_inference_input_mutations=False,
+) -> torch.fx.GraphModule:
+    torch._dynamo.utils.assert_no_fake_params_or_buffers(mod)
+
+    named_parameters = dict(mod.named_parameters(remove_duplicate=False))
+    named_buffers = dict(mod.named_buffers(remove_duplicate=False))
+    params = {
+        **named_parameters,
+        **named_buffers
+    }
+    params_flat, params_spec = pytree.tree_flatten(params)
+    params_flat = tuple(params_flat)
+    params_len = len(params_flat)
+
+    def joint_fw_bw(*args, **kwargs):
+        # forward
+        with stateless._reparametrize_module(
+            mod, pytree.tree_unflatten(args[:params_len], params_spec)
+        ):
+            if isinstance(mod, torch.fx.GraphModule):
+                with fx_traceback.preserve_node_meta(), warnings.catch_warnings():
+                    warnings.filterwarnings(
+                        "ignore", "Anomaly Detection has been enabled."
+                    )
+                    with torch.autograd.detect_anomaly(check_nan=False):
+                        out = Interpreter(mod).run(*args[params_len:], **kwargs)
+            else:
+                out = mod(*args[params_len:], **kwargs)
+
+        # backward
+        assert out[0].requires_grad == True
+        assert out[0].numel() == 1, "Must be a scalar"
+
+        setup_stacktrace_preservation_hooks([out[0].grad_fn])
+
+        with fx_traceback.preserve_node_meta():
+            backward_out = torch.autograd.grad(out[0], args[:len(named_parameters)])
```

Comment:
Obviously, out[0] is a hack here. 

We have 3 options for UX of training mode export. 
1. Assuming out[0] is the loss to backprob with, so user must write the forward function to always return loss as the first return. 
2. Ask user to pass in an index to indicate which output is the loss value.
3. We can require that only one of the output can can require_grad=True. User would have to call .detach on the output that is not the loss. 

Ideas?